Compatibility with Cuda 10.1?

Hi:

Trying to build the most recent OpenCV against **Cuda 10.1**, but failed with the following **ERROR** messages, which I guess is a **opencv~cuda version** compatibility issue?


```console
...../opencv/3rdparty/include/opencl/1.2
....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(176): error: identifier "shfl_up_sync" is undefined
          detected during instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=int]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(220): here

....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(183): error: identifier "shfl_up_sync" is undefined
          detected during instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=int]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(220): here

....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(189): error: identifier "shfl_up_sync" is undefined
          detected during instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=int]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(220): here

....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(176): error: identifier "shfl_up_sync" is undefined
          detected during:
            instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(458): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(cv::cudev::GlobPtr<T>, int, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(584): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(const cv::cudev::GlobPtr<T> &, int, int, cudaStream_t) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(614): here

....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(183): error: identifier "shfl_up_sync" is undefined
          detected during:
            instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(458): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(cv::cudev::GlobPtr<T>, int, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(584): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(const cv::cudev::GlobPtr<T> &, int, int, cudaStream_t) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(614): here

....../opencv_contrib/modules/cudev/include/opencv2/cudev/warp/detail/../../warp/shuffle.hpp(189): error: identifier "shfl_up_sync" is undefined
          detected during:
            instantiation of "T cv::cudev::compatible_shfl_up(T, uint, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(458): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(cv::cudev::GlobPtr<T>, int, int) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(584): here
            instantiation of "void cv::cudev::integral_detail::vertical_pass(const cv::cudev::GlobPtr<T> &, int, int, cudaStream_t) [with T=uint]" 
....../opencv_contrib/modules/cudev/include/opencv2/cudev/grid/detail/integral.hpp(614): here

6 errors detected in the compilation of "/tmp/tmpxft_0000180f_00000000-6_gpu_mat.cpp1.ii".
-- Removing ....../opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
/usr/bin/cmake -E remove ....../opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
CMake Error at cuda_compile_1_generated_gpu_mat.cu.o.Release.cmake:279 (message):
  Error generating file
  ....../opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o


modules/core/CMakeFiles/opencv_core.dir/build.make:497: recipe for target 'modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_gpu_mat.cu.o' failed
make[2]: *** [modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_gpu_mat.cu.o] Error 1
make[2]: Leaving directory '....../opencv/build'
CMakeFiles/Makefile2:2147: recipe for target 'modules/core/CMakeFiles/opencv_core.dir/all' failed
make[1]: *** [modules/core/CMakeFiles/opencv_core.dir/all] Error 2
make[1]: Leaving directory '....../opencv/build'
Makefile:165: recipe for target 'all' failed
make: *** [all] Error 2
```


Cheers
Pei

--- a/modules/cudev/include/opencv2/cudev/warp/shuffle.hpp
+++ b/modules/cudev/include/opencv2/cudev/warp/shuffle.hpp
@@ -157,43 +157,6 @@ CV_CUDEV_SHFL_VEC_INST(double)
 
 // shfl_up
 
-template <typename T>
-__device__ __forceinline__ T compatible_shfl_up(T val, uint delta, int width = warpSize)
-{
-#if __CUDACC_VER_MAJOR__ < 9
-
-    return shfl_up(val, delta, width);
-
-#else // __CUDACC_VER_MAJOR__ < 9
-
-#if CV_CUDEV_ARCH >= 700
-    return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
-#else
-    const int block_size = Block::blockSize();
-    const int residual = block_size & (warpSize - 1);
-
-    if (0 == residual)
-        return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
-    else
-    {
-        const int n_warps = divUp(block_size, warpSize);
-        const int warp_id = Warp::warpId();
-
-        if (warp_id < n_warps - 1)
-            return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
-        else
-        {
-            // We are at the last threads of a block whose number of threads
-            // is not a multiple of the warp size
-            uint mask = (1LU << residual) - 1;
-            return shfl_up_sync(mask, val, delta, width);
-        }
-    }
-#endif
-
-#endif // __CUDACC_VER_MAJOR__ < 9
-}
-
 #if __CUDACC_VER_MAJOR__ >= 9
 
 template <typename T>
@@ -299,6 +262,43 @@ CV_CUDEV_SHFL_UP_VEC_INST(double)
 #undef CV_CUDEV_SHFL_UP_VEC_INST
 
 #endif
+
+template <typename T>
+__device__ __forceinline__ T compatible_shfl_up(T val, uint delta, int width = warpSize)
+{
+#if __CUDACC_VER_MAJOR__ < 9
+
+    return shfl_up(val, delta, width);
+
+#else // __CUDACC_VER_MAJOR__ < 9
+
+#if CV_CUDEV_ARCH >= 700
+    return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
+#else
+    const int block_size = Block::blockSize();
+    const int residual = block_size & (warpSize - 1);
+
+    if (0 == residual)
+        return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
+    else
+    {
+        const int n_warps = divUp(block_size, warpSize);
+        const int warp_id = Warp::warpId();
+
+        if (warp_id < n_warps - 1)
+            return shfl_up_sync(0xFFFFFFFFU, val, delta, width);
+        else
+        {
+            // We are at the last threads of a block whose number of threads
+            // is not a multiple of the warp size
+            uint mask = (1LU << residual) - 1;
+            return shfl_up_sync(mask, val, delta, width);
+        }
+    }
+#endif
+
+#endif // __CUDACC_VER_MAJOR__ < 9
+}
 
 // shfl_down
 

